### PR TITLE
Implement generic CLI query connection end

### PR DESCRIPTION
**Pull request by thaodt**
Part of  #393.

## Description

Following the guideline in issue description to do it.

This PR is still in-draft but I have 1 question for you @soareschen. I'm getting this error:
```bash
error[E0308]: mismatched types
  --> crates/cli/cli-components/src/impls/commands/queries/connection_end.rs:88:31
   |
49 | impl<App, Args, Build, Chain, Counterparty> CommandRunner<App, Args>
   |                        -----  ------------ expected type parameter
   |                        |
   |                        found type parameter
...
88 |         Ok(app.produce_output(connection_end))
   |                -------------- ^^^^^^^^^^^^^^ expected type parameter `Counterparty`, found type parameter `Chain`
   |                |
   |                arguments to this method are incorrect
   |
   = note: expected associated type `<Counterparty as hermes_relayer_components::chain::traits::types::connection::HasConnectionEndType<Chain>>::ConnectionEnd`
              found associated type `<Chain as hermes_relayer_components::chain::traits::types::connection::HasConnectionEndType<Counterparty>>::ConnectionEnd`
   = note: a type parameter was expected, but a different one was found; you might be missing a type parameter or trait bound
   = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters
   = note: an associated type was expected, but a different one was found
note: method defined here
  --> crates/cli/cli-components/src/traits/output.rs:10:8
   |
10 |     fn produce_output(&self, value: Value) -> Self::Output;
   |        ^^^^^^^^^^^^^^
```

I guess it relates to one of traits I used in `Chain` & `Counterparty` when implementing for `RunQueryConnectionEndCommand` but I still didn't figure out what's wrong in my implementation. Can you please point it out for me? Thanks.

### CODE REVIEW

File: crates/cli/cli-components/src/impls/commands/queries/connection_end.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        help = "Height of the state to query. Leave unspecified for latest height."
+    )]
+    height: Option<String>,
+}
+
+impl<App, Args, Build, Chain, Counterparty> CommandRunner<App, Args>
+    for RunQueryConnectionEndCommand
+where
+    App: CanLoadBuilder<Builder = Build>
+        + CanProduceOutput<Counterparty::ConnectionEnd>
```
Comment (soareschen) [inline crates/cli/cli-components/src/impls/commands/queries/connection_end.rs:53]: The connection end returned from `CanQueryConnectionEnd` is of type `Chain::ConnectionEnd`. So you need to use that to produce the CLI output.

File: crates/cli/cli-components/src/impls/commands/queries/connection_end.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        + CanParseArg<Args, symbol!("chain_id"), Parsed = Chain::ChainId>
+        + CanParseArg<Args, symbol!("connection_id"), Parsed = Chain::ConnectionId>
+        + CanParseArg<Args, symbol!("height"), Parsed = Option<Chain::Height>>
+        + CanRaiseError<Build::Error>
+        + CanRaiseError<Chain::Error>
+        + CanRaiseError<String>,
+    Build: CanBuildChain<0, Chain = Chain> + HasChainTypeAt<1, Chain = Counterparty>,
+    Chain: CanQueryChainHeight + CanQueryConnectionEnd<Counterparty>,
+    Counterparty: HasHeightType + HasChainId + HasConnectionEndType<Chain>,
```
Comment (soareschen) [inline crates/cli/cli-components/src/impls/commands/queries/connection_end.rs:62]: All the constraints for `Counterparty` are not really needed, thus can be removed.

File: crates/cli/cli-components/src/impls/commands/queries/connection_end.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    for RunQueryConnectionEndCommand
+where
+    App: CanLoadBuilder<Builder = Build>
+        + CanProduceOutput<Counterparty::ConnectionEnd>
+        + CanParseArg<Args, symbol!("chain_id"), Parsed = Chain::ChainId>
+        + CanParseArg<Args, symbol!("connection_id"), Parsed = Chain::ConnectionId>
+        + CanParseArg<Args, symbol!("height"), Parsed = Option<Chain::Height>>
+        + CanRaiseError<Build::Error>
+        + CanRaiseError<Chain::Error>
+        + CanRaiseError<String>,
```
Comment (soareschen) [inline crates/cli/cli-components/src/impls/commands/queries/connection_end.rs:59]: We didn't raise any string error in this implementation. So the constraint can be removed.

File: crates/cli/cli-components/src/impls/commands/queries/connection_end.rs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+impl<App, Args, Build, Chain, Counterparty> CommandRunner<App, Args>
+    for RunQueryConnectionEndCommand
+where
+    App: CanLoadBuilder<Builder = Build>
+        + CanProduceOutput<Counterparty::ConnectionEnd>
+        + CanParseArg<Args, symbol!("chain_id"), Parsed = Chain::ChainId>
+        + CanParseArg<Args, symbol!("connection_id"), Parsed = Chain::ConnectionId>
+        + CanParseArg<Args, symbol!("height"), Parsed = Option<Chain::Height>>
```
Comment (soareschen) [inline crates/cli/cli-components/src/impls/commands/queries/connection_end.rs:56]: We need to add the following wiring to [`HermesParserComponents`](https://github.com/informalsystems/hermes-sdk/blob/668e55080bef8e13ffffbeceba49d5be74de0ff8/crates/cli/cli/src/contexts/app.rs#L145) to enable the parsing of the given fields.

```
        (QueryConnectionEndArgs, symbol!("chain_id")): ParseFromString<ChainId>,
        (QueryConnectionEndArgs, symbol!("connection_id")): ParseFromString<ConnectionId>,
        (QueryConnectionEndArgs, symbol!("height")): ParseFromOptionalString<Height>,
```

File: crates/cli/cli/src/commands/query/mod.rs [hunk 5]
```diff
@@ -57,7 +54,7 @@ impl CommandRunner<HermesApp> for QueryCommands {
         match self {
             Self::Client(cmd) => app.run_command(cmd).await,
             Self::Clients(cmd) => cmd.run(app).await,
-            Self::Connection(cmd) => cmd.run(app).await,
+            Self::Connection(cmd) => app.run_command(cmd).await,
```
Comment (soareschen) [inline crates/cli/cli/src/commands/query/mod.rs:57]: We need to add the following wiring to [`HermesCommandRunnerComponents`](https://github.com/informalsystems/hermes-sdk/blob/668e55080bef8e13ffffbeceba49d5be74de0ff8/crates/cli/cli/src/contexts/app.rs#L161) to forward the subcommand to the corresponding runners:

```
        QueryConnectionSubCommand: RunQueryConnectionSubCommand,
        QueryConnectionEndArgs: RunQueryConnectionEndCommand,
```

File: crates/cli/cli-components/src/impls/commands/queries/connection_end.rs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+    #[clap(
+        long = "height",
+        value_name = "HEIGHT",
+        help = "Height of the state to query. Leave unspecified for latest height."
+    )]
+    height: Option<String>,
+}
+
+impl<App, Args, Build, Chain, Counterparty> CommandRunner<App, Args>
```
Comment (soareschen) [inline crates/cli/cli-components/src/impls/commands/queries/connection_end.rs:49]: The following constraints should be added to [`CanUseHermesApp`](https://github.com/informalsystems/hermes-sdk/blob/668e55080bef8e13ffffbeceba49d5be74de0ff8/crates/cli/cli/src/contexts/app.rs#L232) to check that the `HermesApp` context indeed implements `CommandRunner` for the corresponding command args:

```
    + CanRunCommand<QueryConnectionSubCommand>
    + CanRunCommand<QueryConnectionEndArgs>
```

File: crates/cli/cli/src/commands/query/mod.rs [hunk 7]
```diff
@@ -1,6 +1,3 @@
-mod connection;
```
Comment (soareschen) [inline crates/cli/cli/src/commands/query/mod.rs:1]: The corresponding files at [`commands/query/connection`](https://github.com/informalsystems/hermes-sdk/tree/668e55080bef8e13ffffbeceba49d5be74de0ff8/crates/cli/cli/src/commands/query/connection) should be removed.

**Review by soareschen — CHANGES_REQUESTED** [review]
Thanks for your contribution! I have added notes on how you can fix the errors.

File: crates/cli/cli/src/commands/query/mod.rs (hunk 5 above)
Comment (thaodt) [reply crates/cli/cli/src/commands/query/mod.rs:58]: @soareschen Im getting error 
```shell
error[E0277]: the trait bound `RunQueryConnectionSubCommand: hermes_cli_components::traits::command::CommandRunner<HermesApp, QueryConnectionSubCommand>` is not satisfied
  --> crates/cli/cli/src/commands/query/mod.rs:58:54
   |
58 |             Self::Connection(cmd) => app.run_command(cmd).await,
   |                                          ----------- ^^^ the trait `hermes_cli_components::traits::command::CommandRunner<HermesApp, QueryConnectionSubCommand>` is not implemented for `RunQueryConnectionSubCommand`, which is required by `HermesApp: CanRunCommand<_>`
   |                                          |
   |                                          required by a bound introduced by this call
   |
   = help: the trait `hermes_cli_components::traits::command::CommandRunner<App, QueryConnectionSubCommand>` is implemented for `RunQueryConnectionSubCommand`
   = note: required for `UseDelegate<HermesCommandRunnerComponents>` to implement `hermes_cli_components::traits::command::CommandRunner<HermesApp, QueryConnectionSubCommand>`
   = note: 1 redundant requirement hidden
   = note: required for `HermesAppComponents` to implement `hermes_cli_components::traits::command::CommandRunner<HermesApp, QueryConnectionSubCommand>`
   = note: required for `HermesApp` to implement `CanRunCommand<QueryConnectionSubCommand>`

error[E0277]: the trait bound `UseDelegate<HermesCommandRunnerComponents>: hermes_cli_components::traits::command::CommandRunner<HermesApp, QueryConnectionSubCommand>` is not satisfied
  --> crates/cli/cli/src/commands/query/mod.rs:58:38
   |
58 |             Self::Connection(cmd) => app.run_command(cmd).await,
   |                                      ^^^^^^^^^^^^^^^^^^^^ the trait `hermes_cli_components::traits::command::CommandRunner<HermesApp, QueryConnectionSubCommand>` is not implemented for `UseDelegate<HermesCommandRunnerComponents>`, which is required by `HermesAppComponents: hermes_cli_components::traits::command::CommandRunner<HermesApp, QueryConnectionSubCommand>`
   |
   = help: the trait `hermes_cli_components::traits::command::CommandRunner<App, Args>` is implemented for `UseDelegate<Components>`
   = note: required for `HermesAppComponents` to implement `hermes_cli_components::traits::command::CommandRunner<HermesApp, QueryConnectionSubCommand>`
note: required by a bound in `<App as CanRunCommand<Args>>`
  --> /home/thaodinh/projects/oss/informalsystems/hermes-sdk/crates/cli/cli-components/src/traits/command.rs:6:1
   |
6  | #[derive_component(CommandRunnerComponent, CommandRunner<App>)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `<App as CanRunCommand<Args>>`
   = note: this error originates in the attribute macro `derive_component` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `UseDelegate<HermesCommandRunnerComponents>: hermes_cli_components::traits::command::CommandRunner<HermesApp, QueryConnectionSubCommand>` is not satisfied
  --> crates/cli/cli/src/commands/query/mod.rs:58:59
   |
58 |             Self::Connection(cmd) => app.run_command(cmd).await,
   |                                                           ^^^^^ the trait `hermes_cli_components::traits::command::CommandRunner<HermesApp, QueryConnectionSubCommand>` is not implemented for `UseDelegate<HermesCommandRunnerComponents>`, which is required by `HermesAppComponents: hermes_cli_components::traits::command::CommandRunner<HermesApp, QueryConnectionSubCommand>`
   |
   = help: the trait `hermes_cli_components::traits::command::CommandRunner<App, Args>` is implemented for `UseDelegate<Components>`
   = note: required for `HermesAppComponents` to implement `hermes_cli_components::traits::command::CommandRunner<HermesApp, QueryConnectionSubCommand>`
 ```
while I already added all necessary constraints, also I tried to expand macros to check the code generated by cgp, but I cannot find out what actually the problem is. Could you please shed some light?

Comment (thaodt) [reply crates/cli/cli/src/commands/query/mod.rs:58]: thank for your guidance @soareschen to help me fix this.

**Comment by thaodt** [comment]
@soareschen I finished the PR, hopefully we can get it merged. ^^

**Comment by soareschen** [comment]
Merged now. Thank you for your contribution!

**Pull request merged**